docs/2019-05-23-Unification-Modulo-Equations.md

```diff
@@ -0,0 +1,163 @@
+constructor (actually, it only needs to be injective), then
+```
+⌈C(φ_1, …, φ_n) ∧ C(ψ_1, …, ψ_n)⌉=⌈φ_1∧ψ_1⌉∧…∧⌈φ_n∧ψ_n⌉
+```
+
+However, if `f` and `g` are random symbols, without any properties, then
+`⌈f(φ_1, …, φ_n) ∧ g(ψ_1, …, ψ_m)⌉` cannot be reduced to a nicer form.
+
+If `f` is, say, an associative binary operation with neutral element (i.e. the
+concatenation for a list), then we have a reasonable intuition about how to
+unify it, but I don’t know of any written text about that.
+
+Problem
+-------
+
+We are trying to extend unification to sorts that do not have constructors, but
+which have something resembling constructors. An example is a list, which has:
+`unit` - the empty list.
+`elem` - takes an element of a different sort and produces a single element list.
+`concat` - concatenates two lists.
+Heating and cooling could be another example.
+
+Proposed Solution
+-----------------
+
+The first attempt to solve this would be to define some sort of
+no-junk/no-confusion axiom
+for that sort (e.g. any list can be built with `unit`, `elem` and `concat`),
+and then expand the two terms in all possible ways using the available
+equations, then match the two terms at the top as long as we stay within
+the same sort, then use unification for the remainders.
+
+However, this has a few problems. First, there is no way of proving that
+this is the most general unifier, i.e. there is no way of showing that we
+truly expanded the term in all possible ways. Second, this would be really
+inefficient.
+
+### Actual proposal
+
+The second attempt, however, would be to define some unification axioms (I
+think that they are, indeed, axioms, but they could be theorems if we find
+a way to infer them), specifying how to solve certain unification cases
+in a way that can be used almost directly by the Haskell backend (some
+changes needed).
+
+These axioms would allow us to reduce the hard to handle
+unification cases to unambiguous ones (e.g. reduce AC unification to
+constructor-based unification).
+
+See the use-cases below for examples and explanations.
+
+We could write these in a way that we don't duplicate unification cases.
+Then we could apply these as simplification axioms, defining our unification
+strategy in k. When we need performance, we could implement the ones that are
+used more frequently in Haskell.
+
+Or, if that ever becomes desirable (e.g. if we want it to make proof
+certificates easier to generate), we could give up on performance and define
+all simplification cases as axioms/theorems.
+
+Also, we should probably use set variables, to avoid the large number of
+checks needed when working with normal variables (though, of course, we
+need to make sure that the axioms still hold).
+
+Use-case: lists
+---------------
+
+For lists, we could have the following axioms
+```
+⌈elem(x)∧unit⌉ = ⊥
+⌈elem(x)∧elem(y)⌉ = ⌈x∧y⌉
+⌈concat(x,y)∧unit⌉ = ⌈x∧unit⌉ ∧ ⌈y∧unit⌉
```

Comment:
`⌈concat(x,y)∧unit⌉` can directly equal to `⊥` right?

Reply:
Not if x and y are unit.

Reply:
No. One can take `x=unit` and `y=unit`. Suggestion: use `x,y,z,...` for variables of `Nat` and `l1,l2,...` for variables of `List`.

Reply:
yes...you're right. empty list is not bottom

Reply:
Thanks, I replaced the list variables.